File: app/src/main/java/com/rezolve/sdk_sample/MainActivity.java
package com.rezolve.sdk_sample;

import android.Manifest;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.nabinbhandari.android.permissions.PermissionHandler;
import com.nabinbhandari.android.permissions.Permissions;
import com.rezolve.sdk.RezolveInterface;
import com.rezolve.sdk.RezolveSDK;
import com.rezolve.sdk.RezolveSession;
import com.rezolve.sdk.location.LocationDependencyProvider;
import com.rezolve.sdk.model.network.RezolveError;
import com.rezolve.sdk.model.shop.Product;
import com.rezolve.sdk_sample.model.AuthenticationResponse;
import com.rezolve.sdk_sample.providers.AuthenticationServiceProvider;
import com.rezolve.sdk_sample.providers.SdkProvider;
import com.rezolve.sdk_sample.remote.ScanActivityRemote;
import com.rezolve.sdk_sample.services.AuthenticationService;
import com.rezolve.sdk_sample.services.callbacks.AuthenticationCallback;
import com.rezolve.sdk_sample.utils.DeviceUtils;
import com.rezolve.sdk_sample.utils.DialogUtils;
import com.rezolve.sdk_sample.utils.NotificationUtil;
import com.rezolve.sdk_sample.utils.ProductUtils;

import static com.rezolve.sdk_sample.utils.NotificationUtil.isLaunchedFromNotification;

public class MainActivity extends AppCompatActivity implements MainNavigator {

    private String deviceId;

    private AuthenticationService authenticationService = AuthenticationServiceProvider.getAuthenticationService();
    private RezolveSDK rezolveSDK = SdkProvider.getInstance().getSDK();

    private Button localScanBtn, remoteScanBtn;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        loginUser();
        prepareNavigationButtons();
    }

    @Override
    protected void onResume() {
        super.onResume();

        String[] locationPermissions = { Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION };
        Permissions.check(this, locationPermissions, null, null, new PermissionHandler() {
            @Override
            public void onGranted() {
                LocationDependencyProvider.locationProvider().start();
            }
        });
    }

    private void prepareNavigationButtons() {
        localScanBtn = findViewById(R.id.local_scan_activity_btn);
        remoteScanBtn = findViewById(R.id.remote_scan_activity_btn);
        localScanBtn.setOnClickListener(view -> navigateToScanView());
        remoteScanBtn.setOnClickListener(view -> navigateToRemoteScanView());
    }

    private void loginUser() {
        deviceId = DeviceUtils.getDeviceId(this);
        authenticationService.login(BuildConfig.DEMO_AUTH_USER, BuildConfig.DEMO_AUTH_PASSWORD, deviceId, new AuthenticationCallback() {
            @Override
            public void onLoginSuccess(AuthenticationResponse response) {
                createSession(response);
            }

            @Override
            public void onLoginFailure(String message) {
                DialogUtils.showError(MainActivity.this, message);
            }
        });
    }

    private void createSession(AuthenticationResponse response) {

        rezolveSDK.setAuthToken(response.getToken());
        rezolveSDK.setDeviceIdHeader(deviceId);

        rezolveSDK.createSession(response.getToken(), response.getEntityId(), response.getPartnerId(), null, new RezolveInterface() {
            @Override
            public void onInitializationSuccess(RezolveSession rezolveSession, String partnerId, String entityId) {
                if(isLaunchedFromNotification(MainActivity.this)) {
                    NotificationUtil.launch(getIntent(), MainActivity.this);
                } else {
                    showNavigationButtons();
                }
            }

            @Override
            public void onInitializationFailure(@NonNull RezolveError rezolveError) {
                DialogUtils.showError(MainActivity.this, rezolveError.getMessage());
            }
        });
    }

    private void showNavigationButtons() {
        localScanBtn.setVisibility(View.VISIBLE);
        remoteScanBtn.setVisibility(View.VISIBLE);
        findViewById(R.id.authenticationStatusTextView).setVisibility(View.GONE);
    }

    private void navigateToScanView() {
        Intent intent = new Intent(MainActivity.this, ScanActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        startActivity(intent);
    }

    private void navigateToRemoteScanView() {
        Intent intent = new Intent(MainActivity.this, ScanActivityRemote.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        startActivity(intent);
    }

    public void navigateToProductDetails(@NonNull Product product) {
        // We want to ensure that BGL is stopped even after fresh launch from notification

        Intent intent = new Intent(MainActivity.this, ProductDetailsActivity.class);
        Bundle bundle = ProductUtils.toBundle(product);
        intent.putExtras(bundle);
        startActivity(intent);
    }
}

File: app_location_triggers/src/main/java/com/rezolve/smart_triggers/MainActivity.kt
package com.rezolve.smart_triggers

import android.Manifest
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.PermissionChecker
import com.rezolve.rxp.client.APIResult
import com.rezolve.rxp.client.data.model.Geofences
import com.rezolve.rxp.client.data.model.Tag
import com.rezolve.rxp.enums.CoordinateSystem
import com.rezolve.rxp.enums.MyAreaFilter
import com.rezolve.rxp.enums.State
import com.rezolve.rxp.sdk.RxpSdkProvider
import com.rezolve.sdk.RezolveInterface
import com.rezolve.sdk.RezolveSession
import com.rezolve.sdk.location.LocationHelper
import com.rezolve.sdk.location.LocationUpdateListener
import com.rezolve.sdk.location.LocationWrapper
import com.rezolve.sdk.location.google.LocationProviderFused
import com.rezolve.sdk.model.network.RezolveError
import com.rezolve.shared.authentication.AuthenticationCallback
import com.rezolve.shared.authentication.AuthenticationResponse
import com.rezolve.shared.authentication.AuthenticationServiceProvider
import com.rezolve.shared.utils.DeviceUtils
import com.rezolve.shared.utils.DialogUtils
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import sb.rezolve.app.sandbox.BuildConfig
import sb.rezolve.app.sandbox.R

class MainActivity : AppCompatActivity() {

    private lateinit var deviceId: String

    private val REQUEST_CODE_LOCATION = 10203
    private val REQUEST_CODE_BACKGROUND_LOCATION = 10204
    private val BASE_PERMISSIONS = arrayOf(
        Manifest.permission.ACCESS_FINE_LOCATION,
        Manifest.permission.ACCESS_COARSE_LOCATION,
        Manifest.permission.POST_NOTIFICATIONS // required to show notifications on Android 33+
    )

    @RequiresApi(Build.VERSION_CODES.M)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        if (LocationHelper.isLocationPermissionGranted(this).not()
            || PermissionChecker.checkSelfPermission(this, Manifest.permission.POST_NOTIFICATIONS) != PermissionChecker.PERMISSION_GRANTED
            || PermissionChecker.checkSelfPermission(this, Manifest.permission.ACCESS_BACKGROUND_LOCATION) != PermissionChecker.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, BASE_PERMISSIONS, REQUEST_CODE_LOCATION)
        } else {
            loginUser()
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {

        when(requestCode) {
            REQUEST_CODE_LOCATION -> {
                if (grantResults.size == BASE_PERMISSIONS.size && grantResults.contains(PackageManager.PERMISSION_DENIED).not()) {
                    /** required to register geofences on android 29+
                     * AND has to be requested AFTER basic location permissions were granted
                     * https://developer.android.com/develop/sensors-and-location/location/permissions#request-background-location
                     */
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q && PermissionChecker.checkSelfPermission(this, Manifest.permission.ACCESS_BACKGROUND_LOCATION) != PermissionChecker.PERMISSION_GRANTED) {
                        ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.ACCESS_BACKGROUND_LOCATION), REQUEST_CODE_BACKGROUND_LOCATION)
                    } else {
                        loginUser()
                    }
                } else {
                    Toast.makeText(this, getString(R.string.missing_location_permission), Toast.LENGTH_LONG).show()
                }
            }
            REQUEST_CODE_BACKGROUND_LOCATION -> {
                if (grantResults.contains(PackageManager.PERMISSION_DENIED).not()) {
                    loginUser()
                } else {
                    Toast.makeText(this, getString(R.string.missing_location_permission), Toast.LENGTH_LONG).show()
                }
            }
        }

        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    private fun startLocationUpdates() {

        val locationProviderFused = LocationProviderFused.create(this)

        LocationHelper.getInstance(this).addLocationListener(object : LocationUpdateListener {
            override fun onLocationChanged(location: LocationWrapper?) {
                Log.d(TAG, "location update: ${location?.entityToJson()}")
                location?.let {
                    CoroutineScope(Dispatchers.IO).launch {
                        val updateTrackingApiResult: APIResult<Geofences> = RxpSdkProvider.sdk.rxpClient.updateTracking(
                            location.rezolveLocation.latitude.toFloat(),
                            location.rezolveLocation.longitude.toFloat(),
                            2000,
                            CoordinateSystem.WGS84
                        )
                    }
                }
            }
        })

        locationProviderFused.start()
    }

    private fun loginUser() {
        deviceId = DeviceUtils.getDeviceId(this)
        AuthenticationServiceProvider.getAuthenticationService().login(
            BuildConfig.DEMO_AUTH_USER,
            BuildConfig.DEMO_AUTH_PASSWORD,
            deviceId,
            object : AuthenticationCallback {
                override fun onLoginSuccess(response: AuthenticationResponse) {
                    createSession(response)
                }

                override fun onLoginFailure(message: String) {
                    DialogUtils.showError(this@MainActivity, message)
                }
            })
    }

    private fun createSession(response: AuthenticationResponse) {
        SdkProvider.rezolveSdk.apply {
            setAuthToken(response.token)
            setDeviceIdHeader(deviceId)

            SdkProvider.tokenHolder.accessToken = response.token

            createSession(
                response.token,
                response.entityId,
                response.partnerId,
                null,
                object : RezolveInterface {
                    override fun onInitializationSuccess(
                        rezolveSession: RezolveSession,
                        partnerId: String,
                        entityId: String
                    ) {
                        Log.d(TAG, "RezolveSDK initalization success")
                        startLocationUpdates()

                        CoroutineScope(Dispatchers.IO).launch {
                            rxpCheckIn()
                        }
                    }

                    override fun onInitializationFailure(rezolveError: RezolveError) {
                        DialogUtils.showError(this@MainActivity, rezolveError.message)
                    }
                })
        }
    }

    private suspend fun rxpCheckIn() {
        SdkProvider.pushNotificationProvider.pushToken.collect {
            val checkInApiResult: APIResult<String> = RxpSdkProvider.sdk.rxpClient.checkIn(it)
            if (checkInApiResult is APIResult.Success) {
                //handle successful check in
                Log.d(AppST.TAG, "RxpSdk check-in success: ${checkInApiResult.result}")
                RxpSdkProvider.sdk.runTagsListWorkerOnce()

                val tagsApiResult = RxpSdkProvider.sdk.rxpClient.listTags()
                if (tagsApiResult is APIResult.Success) {
                    val tags = tagsApiResult.result
                    val newTags = mutableListOf<Tag>()
                    tags.forEach { tag ->
                        newTags.add(tag.copy(state = State.ENABLED)) // update tags to user preferences
                    }
                    RxpSdkProvider.sdk.rxpClient.updateTags(newTags)
                }
            } else if (checkInApiResult is APIResult.Error) {
                //handle error
                withContext(Dispatchers.Main) {
                    DialogUtils.showError(this@MainActivity, checkInApiResult.message)
                }
            }
        }
    }

    // returns paginated list of engagements in selected proximity
    private fun getNearbyEngagements() {
        LocationHelper.getInstance(this).lastKnownLocation?.let {
            val result = RxpSdkProvider.sdk.rxpClient.getMyArea(
                latitude = it.rezolveLocation.latitude.toFloat(),
                longitude = it.rezolveLocation.longitude.toFloat(),
                distance = 2000,
                coordinateSystem = CoordinateSystem.WGS84,
                filter = MyAreaFilter.ALL, // use MyAreaFilter.MY to get nearby engagements filtered by user's interests
                limit = 100,
                offset = 0
            )
        }
    }

    companion object {
        const val TAG = "MainActivity"
    }
}
